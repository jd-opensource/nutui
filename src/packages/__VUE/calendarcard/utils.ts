import Utils from '@/packages/utils/date'
import { CalendarCardDay, CalendarCardMonth, CalendarCardValue } from './types'

export const convertDateToDay = (date: Date) => {
  return date
    ? {
        year: date.getFullYear(),
        month: date.getMonth() + 1,
        date: date.getDate()
      }
    : null
}

export const convertDayToDate = (day: CalendarCardDay) => {
  return day ? new Date(day.year, day.month - 1, day.date) : null
}

export const valueToRange = (val?: CalendarCardValue) => {
  if (Array.isArray(val)) {
    return val.map((date: Date) => {
      return convertDateToDay(date)
    })
  }
  return val ? [convertDateToDay(val)] : []
}

export const rangeTovalue = (range?: CalendarCardDay[]) => {
  if (Array.isArray(range)) {
    return range.map((day: CalendarCardDay) => {
      return convertDayToDate(day)
    })
  }
  return range ? [convertDayToDate(range)] : []
}

/**
 * 获取当月面板中前一个月的日期数据
 */
export const getPrevMonthDays = (year: number, month: number, firstDayOfWeek: number) => {
  let prevMonth = month - 1
  let prevYear = year
  if (prevMonth <= 0) {
    prevMonth = 12
    prevYear -= 1
  }
  let days = Utils.getMonthPreDay(year, month)
  days -= firstDayOfWeek
  if (days >= 7) {
    days -= 7
  }

  const preDates = Utils.getMonthDays(`${prevYear}`, `${prevMonth}`)
  const months = Array.from(Array(preDates), (_, k) => {
    return {
      type: 'prev',
      year: prevYear,
      month: prevMonth,
      date: k + 1
    } as CalendarCardDay
  })
  return months.slice(preDates - days)
}

/**
 * 获取当前月的日期数据
 */
export const getCurrentMonthDays = (year: number, month: number) => {
  const days = Utils.getMonthDays(`${year}`, `${month}`)
  return Array.from(Array(days), (_, k) => {
    return {
      type: 'current',
      year,
      month,
      date: k + 1
    } as CalendarCardDay
  })
}

/**
 * 根据日期获取当前周的起始日期
 */
export const getCurrentWeekDays = (day: CalendarCardDay, firstDayOfWeek: number): CalendarCardDay[] => {
  const current = new Date(day.year, day.month - 1, day.date)
  const count = (current.getDay() + 7 - firstDayOfWeek) % 7
  return [
    convertDateToDay(new Date(current.getTime() - 24 * 60 * 60 * 1000 * count)) as CalendarCardDay,
    convertDateToDay(new Date(current.getTime() + 24 * 60 * 60 * 1000 * (6 - count))) as CalendarCardDay
  ]
}

/**
 * 获取 month 月份数据
 * @param month
 * @param firstDayOfWeek
 * @returns
 */
export const getDays = (month: CalendarCardMonth, firstDayOfWeek: number) => {
  const y = month.year
  const m = month.month
  const days = [...getPrevMonthDays(y, m, firstDayOfWeek), ...getCurrentMonthDays(y, m)] as CalendarCardDay[]
  const size = days.length
  const yearOfNextMonth = month.month === 12 ? month.year + 1 : month.year
  const monthOfNextMonth = month.month === 12 ? 1 : month.month + 1
  // 补全 6 行 7 列视图
  for (let i = 1; i <= 42 - size; i++) {
    days.push({
      type: 'next',
      year: yearOfNextMonth,
      month: monthOfNextMonth,
      date: i
    })
  }
  return days
}

/**
 * 比较两个日期
 * @param day1
 * @param day2
 * @returns
 */
export const compareDay = (day1: CalendarCardDay, day2: CalendarCardDay) => {
  if (day1 && day2) {
    if (day1.year === day2.year) {
      if (day1.month === day2.month) {
        return day1.date - day2.date
      }
      return day1.month - day2.month
    }
    return day1.year - day2.year
  }
}

export const isSameDay = (day1: CalendarCardDay, day2: CalendarCardDay) => {
  return compareDay(day1, day2) === 0
}
