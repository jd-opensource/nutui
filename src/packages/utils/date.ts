const Utils = {
  /**
   * 是否为闫年
   * @return {Boolse} true|false
   */
  isLeapYear: function (y: number): boolean {
    return (y % 4 == 0 && y % 100 != 0) || y % 400 == 0;
  },

  /**
   * 返回星期数
   * @return {String}
   */
  getWhatDay: function (year: number, month: number, day: number): string {
    const date = new Date(year + '/' + month + '/' + day);
    const index = date.getDay();
    const dayNames = ['星期日', '星期一', '星期二', '星期三', '星期四', '星期五', '星期六'];
    return dayNames[index];
  },

  /**
   * 返回星期数
   * @return {Number}
   */
  getMonthPreDay: function (year: number, month: number): number {
    const date = new Date(year + '/' + month + '/01');
    let day = date.getDay();
    if (day == 0) {
      day = 7;
    }
    return day;
  },

  /**
   * 返回月份天数
   * @return {Number}
   */
  getMonthDays: function (year: string, month: string): number {
    if (/^0/.test(month)) {
      month = month.split('')[1];
    }
    return ([0, 31, this.isLeapYear(Number(year)) ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31] as number[])[
      month as any
    ];
  },

  /**
   * 补齐数字位数
   * @return {string}
   */
  getNumTwoBit: function (n: number): string {
    n = Number(n);
    return (n > 9 ? '' : '0') + n;
  },

  /**
   * 日期对象转成字符串
   * @return {string}
   */
  date2Str: function (date: Date, split?: string): string {
    split = split || '-';
    const y = date.getFullYear();
    const m = this.getNumTwoBit(date.getMonth() + 1);
    const d = this.getNumTwoBit(date.getDate());
    return [y, m, d].join(split);
  },

  /**
   * 返回日期格式字符串
   * @param {Number} 0返回今天的日期、1返回明天的日期，2返回后天得日期，依次类推
   * @return {string} '2014-12-31'
   */
  getDay: function (i: number): string {
    i = i || 0;
    let date = new Date();
    const diff = i * (1000 * 60 * 60 * 24);
    date = new Date(date.getTime() + diff);
    return this.date2Str(date);
  },

  /**
   * 时间比较
   * @return {Boolean}
   */
  compareDate: function (date1: string, date2: string): boolean {
    const startTime = new Date(date1.replace('-', '/').replace('-', '/'));
    const endTime = new Date(date2.replace('-', '/').replace('-', '/'));
    if (startTime >= endTime) {
      return false;
    }
    return true;
  },

  /**
   * 时间是否相等
   * @return {Boolean}
   */
  isEqual: function (date1: string, date2: string): boolean {
    const startTime = new Date(date1).getTime();
    const endTime = new Date(date2).getTime();
    if (startTime == endTime) {
      return true;
    }
    return false;
  },
  getMonthWeek: function (year: string, month: string, date: string, firstDayOfWeek: number = 0): number {
    const dateNow = new Date(Number(year), parseInt(month) - 1, Number(date));
    let w = dateNow.getDay(); //星期数
    let d = dateNow.getDate();
    let remainder = 6 - w;
    if (firstDayOfWeek !== 0) {
      w = w == 0 ? 7 : w;
      remainder = 7 - w;
    }
    return Math.ceil((d + remainder) / 7);
  },
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  getYearWeek: function (year: string, month: string, date: string, firstDayOfWeek: number = 0): number {
    const dateNow = new Date(Number(year), parseInt(month) - 1, Number(date));
    const dateFirst = new Date(Number(year), 0, 1);
    const dataNumber = Math.round((dateNow.valueOf() - dateFirst.valueOf()) / 86400000);
    return Math.ceil((dataNumber + (dateFirst.getDay() + 1 - 1)) / 7);
  },
  getWeekDate: function (year: string, month: string, date: string, firstDayOfWeek: number = 0): string[] {
    const dateNow = new Date(Number(year), parseInt(month) - 1, Number(date));
    const nowTime = dateNow.getTime();
    let day = dateNow.getDay();
    if (firstDayOfWeek === 0) {
      const oneDayTime = 24 * 60 * 60 * 1000;
      //显示周日
      const SundayTime = nowTime - day * oneDayTime; //本周的周日
      //显示周六
      const SaturdayTime = nowTime + (6 - day) * oneDayTime; //本周的周六

      const sunday = this.date2Str(new Date(SundayTime));
      const saturday = this.date2Str(new Date(SaturdayTime));
      return [sunday, saturday];
    } else {
      day = day == 0 ? 7 : day;
      const oneDayTime = 24 * 60 * 60 * 1000;
      //显示周一
      const MondayTime = nowTime - (day - 1) * oneDayTime; //本周的周一
      //显示周日
      const SundayTime = nowTime + (7 - day) * oneDayTime; //本周的周日

      const monday = this.date2Str(new Date(MondayTime));
      const sunday = this.date2Str(new Date(SundayTime));
      return [monday, sunday];
    }
  },
  formatResultDate: function (date: string) {
    let days = [...date.split('-')];
    days[2] = Utils.getNumTwoBit(Number(days[2]));
    days[3] = `${days[0]}-${days[1]}-${days[2]}`;
    days[4] = Utils.getWhatDay(+days[0], +days[1], +days[2]);
    return days;
  }
};

export default Utils;
