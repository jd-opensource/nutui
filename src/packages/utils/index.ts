export * from './useRelation';
export * from './useLocale';
