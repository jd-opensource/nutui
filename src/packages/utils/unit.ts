import { ComponentPublicInstance, nextTick } from 'vue';
import { VueWrapper, DOMWrapper } from '@vue/test-utils';

function getTouch(el: Element | Window, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5
  };
}

export function trigger(
  wrapper: VueWrapper<ComponentPublicInstance<any, any, any>> | DOMWrapper<Element> | Element | Window,
  eventName: string,
  x = 0,
  y = 0,
  options: any = {}
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper;
  const touchList = options.touchList || [getTouch(el, x, y)];

  if (options.x || options.y) {
    touchList.push(getTouch(el, options.x, options.y));
  }

  const event = document.createEvent('CustomEvent');
  event.initCustomEvent(eventName, true, true, {});

  Object.assign(event, {
    clientX: x,
    clientY: y,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList
  });

  el.dispatchEvent(event);

  return nextTick();
}
export async function mockScrollTop(value: number) {
  Object.defineProperty(window, 'scrollTop', { value, writable: true });
  trigger(window, 'scroll');
  return nextTick();
}

export function sleep(delay = 0): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(resolve, delay);
  });
}
