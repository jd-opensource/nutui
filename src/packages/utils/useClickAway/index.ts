import { Ref, isRef, onMounted, onUnmounted } from 'vue';

type TargetType = Element | Ref<Element | undefined>;

export function useClickAway(onClickAway: () => void, target: TargetType | TargetType[], eventName = 'touchstart') {
  const handler = (event: Event) => {
    const targets = Array.isArray(target) ? target : [target];
    const inTarget = targets.some((item) => {
      const targetElement = isRef(item) ? item.value : item;
      return !targetElement || targetElement.contains(event.target as Node);
    });
    if (!inTarget) {
      onClickAway();
    }
  };

  onMounted(() => {
    window.addEventListener(eventName, handler, true);
  });

  onUnmounted(() => {
    window.removeEventListener(eventName, handler, true);
  });
}
