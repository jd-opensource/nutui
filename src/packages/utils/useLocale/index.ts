import locale from '@/packages/locale';
import { getPropByPath, isFunction } from '../util';

export const useLocale = (name = '') => {
  return (keyPath: string, ...args: unknown[]) => {
    name = name.toLocaleLowerCase();
    const languages = locale.languages();
    let text = keyPath;
    if (name && name.startsWith('nut')) {
      text = `${name.slice(3)}.${keyPath}`;
    }
    const res = getPropByPath(languages, text) || getPropByPath(languages, keyPath);
    return isFunction(res) ? res(...args) : res;
  };
};
