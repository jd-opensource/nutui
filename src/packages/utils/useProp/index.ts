import { computed, getCurrentInstance } from 'vue';
import type { ComponentPublicInstance, ComputedRef } from 'vue';

export const useProp = <T>(name: string): ComputedRef<T | undefined> => {
  const vm = getCurrentInstance();
  if (!vm) {
    throw new Error('useProp must be called within a setup function');
  }
  const proxy = vm.proxy as ComponentPublicInstance;
  const props = proxy?.$props as Record<string, T>;
  return computed(() => props?.[name]);
};
