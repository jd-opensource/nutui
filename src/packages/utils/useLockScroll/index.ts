import { body } from '@/packages/utils/util'
let count = 0

const CLSNAME = 'nut-overflow-hidden'

export const useLockScroll = (isLock: () => boolean) => {
  const lock = () => {
    if (isLock()) {
      try {
        !count && body.classList.add(CLSNAME)
        count++
      } catch (error) {
        console.warn('[NutUI] <useLockScroll>', error)
      }
    }
  }

  const unlock = () => {
    if (isLock() && count) {
      try {
        count--
        !count && body.classList.remove(CLSNAME)
      } catch (error) {
        console.warn('[NutUI] <unlock>', error)
      }
    }
  }

  return [lock, unlock]
}
