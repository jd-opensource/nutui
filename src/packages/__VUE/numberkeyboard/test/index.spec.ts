import NumberKeyboard from '../index.vue';
import NutPopup from '../../popup/index.vue';
import { config, mount } from '@vue/test-utils';
import { trigger, mockScrollTop } from './../../../utils/unit';

beforeAll(() => {
  config.global.components = {
    NutPopup
  };
});

afterAll(() => {
  config.global.components = {};
});

function clickKey(key: Parameters<typeof trigger>[0]) {
  trigger(key, 'touchstart');
  trigger(key, 'touchend');
}

test('should emit input event after clicking number key', async () => {
  const wrapper = mount(NumberKeyboard, {
    props: {
      visible: true,
      isWrapTeleport: false
    }
  });
  clickKey(wrapper.findAll('.key')[0]);
  expect(wrapper.emitted('input')![0]).toEqual([1]);
  wrapper.unmount();
});
test('should shuffle key order when using random-key prop', async () => {
  const wrapper = mount(NumberKeyboard, {
    props: {
      visible: true,
      isWrapTeleport: false,
      randomKeys: true
    }
  });

  expect(wrapper.html()).toMatchSnapshot();
});
test('should emit delete event after clicking delete key', () => {
  const wrapper = mount(NumberKeyboard, {
    props: {
      visible: true,
      isWrapTeleport: false
    }
  });
  clickKey(wrapper.findAll('.key')[11]);
  expect(wrapper.emitted('delete')).toBeTruthy();
});
test('should emit close event after clicking collapse key', () => {
  const wrapper = mount(NumberKeyboard, {
    props: {
      visible: true,
      isWrapTeleport: false
    }
  });
  clickKey(wrapper.findAll('.key')[9]);
  expect(wrapper.emitted('close')).toBeTruthy();
});

test('should emit close event after clicking close button', () => {
  const wrapper = mount(NumberKeyboard, {
    props: {
      visible: true,
      isWrapTeleport: false,
      type: 'rightColumn'
    }
  });
  clickKey(wrapper.find('.key-board-finish'));
  expect(wrapper.emitted('close')).toBeFalsy();
});

test('should render title and close button correctly', () => {
  const wrapper = mount(NumberKeyboard, {
    props: {
      visible: true,
      isWrapTeleport: false,
      title: '默认键盘'
    }
  });
  let title = wrapper.find('.tit');
  expect(title.html()).toContain('默认键盘');
});
test('should emit "update:modelValue" event after clicking key', () => {
  const wrapper = mount(NumberKeyboard, {
    props: {
      visible: true,
      isWrapTeleport: false,
      maxlength: 6
    }
  });
  const keys = wrapper.findAll('.key');

  clickKey(keys[0]);
  expect(wrapper.emitted('update:value')![0]).toEqual(['1']);

  clickKey(keys[1]);
  expect(wrapper.emitted('update:value')![1]).toEqual(['2']);
});
