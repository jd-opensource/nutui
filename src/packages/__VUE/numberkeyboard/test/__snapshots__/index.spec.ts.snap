// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should shuffle key order when using random-key prop 1`] = `
"<view>
  <transition-stub class=\\"nut-numberkeyboard-overlay\\">
    <view class=\\"nut-overlay\\" style=\\"animation-duration: 0.3s; z-index: 2000;\\"></view>
  </transition-stub>
  <transition-stub>
    <view class=\\"nut-popup popup-bottom\\" style=\\"z-index: 2000; animation-duration: 0.3s;\\">
      <div class=\\"nut-numberkeyboard\\">
        <!--v-if-->
        <div class=\\"number-board-body\\">
          <div class=\\"number-board\\">
            <div class=\\"key-board-wrapper\\">
              <div class=\\"key\\">3
                <!--v-if-->
                <!--v-if-->
              </div>
            </div>
            <div class=\\"key-board-wrapper\\">
              <div class=\\"key\\">2
                <!--v-if-->
                <!--v-if-->
              </div>
            </div>
            <div class=\\"key-board-wrapper\\">
              <div class=\\"key\\">6
                <!--v-if-->
                <!--v-if-->
              </div>
            </div>
            <div class=\\"key-board-wrapper\\">
              <div class=\\"key\\">1
                <!--v-if-->
                <!--v-if-->
              </div>
            </div>
            <div class=\\"key-board-wrapper\\">
              <div class=\\"key\\">5
                <!--v-if-->
                <!--v-if-->
              </div>
            </div>
            <div class=\\"key-board-wrapper\\">
              <div class=\\"key\\">4
                <!--v-if-->
                <!--v-if-->
              </div>
            </div>
            <div class=\\"key-board-wrapper\\">
              <div class=\\"key\\">7
                <!--v-if-->
                <!--v-if-->
              </div>
            </div>
            <div class=\\"key-board-wrapper\\">
              <div class=\\"key\\">9
                <!--v-if-->
                <!--v-if-->
              </div>
            </div>
            <div class=\\"key-board-wrapper\\">
              <div class=\\"key\\">8
                <!--v-if-->
                <!--v-if-->
              </div>
            </div>
            <div class=\\"key-board-wrapper\\">
              <div class=\\"key lock\\">
                <!--v-if--><img src=\\"https://img11.360buyimg.com/imagetools/jfs/t1/146371/38/8485/738/5f606425Eca239740/14f4b4f5f20d8a68.png\\">
                <!--v-if-->
              </div>
            </div>
            <div class=\\"key-board-wrapper\\">
              <div class=\\"key\\">0
                <!--v-if-->
                <!--v-if-->
              </div>
            </div>
            <div class=\\"key-board-wrapper\\">
              <div class=\\"key delete\\">
                <!--v-if-->
                <!--v-if--><img src=\\"https://img11.360buyimg.com/imagetools/jfs/t1/129395/8/12735/2030/5f61ac37E70cab338/fb477dc11f46056c.png\\">
              </div>
            </div>
          </div>
          <!--v-if-->
        </div>
      </div>
      <!--v-if-->
    </view>
  </transition-stub>
</view>"
`;
